### DISCUSSION TITLE
monitoring analog or digital input ESP8266 to show data in domoticz

### DISCUSSION THREAD

**Discussion by Damave** [arendst/Tasmota#10331] [Q&A] [answered]
Hi, How can I use the analog input or GPIO2 as a digital input to monitor the time that my waterpump is on. I have already a setup to monitoring power going to the waterpump ( power of pump is 1500 Watt and 80 meters under ground). This information should be stored in Domoticz or somewhere else. I can't power the pump not more than 120 times a hour. The time that pump is running is at least 10 seconds. I do need to measure it by wifi. Don't need to know the current, only the time it is on.

I have tried the 9.2.0 tasmota lite version, but how can i add the analog signal input and transmit this to domoticz?

**Comment by barbudor** [comment, marked as answer]
You can use the analog input by defining it in the template as GPIO17:
![image](https://user-images.githubusercontent.com/2996042/103415832-272da080-4b84-11eb-8b80-f050c2990435.png)

Some details here : https://tasmota.github.io/docs/Expanding-Tasmota/#analog-input
Also look at the command [AdcParam](https://tasmota.github.io/docs/Commands/#adcparam)

For a digital input, a GPIO should be defined as a Switch.
The most appropriate [SwitchMode](https://tasmota.github.io/docs/Commands/#switchmode) should be `switchmode<x> 15` where <x> is the index of your switch. For example for Switch1 use `SwitchMode1 15`
See more details on Switches in [Buttons and Switches](https://tasmota.github.io/docs/Buttons-and-Switches/)

Cheers

**Comment by Frogmore42** [comment]
It kind of depends on exactly what you want.  I have a similar question for my hay feeder.  I want to know how long the hay feeder motor is running.  My case is different, since the maximum time the motor runs is less than 10 seconds.  A well pump of the size you have should be running for a minimum of 60 seconds and preferably for 120 seconds to allow for proper cooling.  You can use a rule to publish a mqtt message when the power is greater than some amount and a different one when it is below that.  Domoticz should be able to understand that as the pump is on and give you what you need.  There is an example of this with the INA219 in the docs, but it should work with whatever you are using to measure the power, as long as it is running Tasmota.

Now, if you want to have an actual input measure something do what @barbudor suggested.

**Reply by barbudor** [reply]
Energy monitoring is intended to measure energy, not to detect operating status, especially short time.
I am also using sockets with Energy monitoring to track record of usage such as TV, computers, etc... So far not using the energy measurment just the operating status. It works fine with the resolution of teleperiod, improved by powerdelta.

But this is not adapted for short events. Even if using rules you can make it work more or less, energy measurments sensors such as HW8012 have a certain latency to detect power up or power down.
To only detect on/off of short events, I would use a totally different approach with an hardware based sensor that will provide a digital input to Tasmota as a switch.
A hall effect based current sensor with a comparator providing 0/1 would be much more efficient.

**Comment by sfromis** [comment]
An easy way of monitoring the pump (I'm assuming mains AC) is if you can add an energy monitoring plug/switch to the circuit. While you're not interested in energy consumption, seeing energy used at all is a pretty good indication of the pump being on. You can use PowerDelta to get messages when it turns on/off, instead of just getting numbers by TelePeriod.

**Comment by Damave** [comment]
It seems to be equal. I can run the pump 24/7. It's hanging in the well, so heat it not the problem. There is a capacitor inside the pump, that can be the reason. I have https://www.italtecnica.com/en/product/pressoflussostati/brio-tank to control the pump. 

I can measure the current going into the pump, by a current transformer 3 winding to 100 windings. I can get now a current(voltage over a resistor) out of the 100 windings, that I can measure with a esp8266. Now the big question. I would like to have a duty factor ( power on per 5 minuts) and the amount of startups.

I can see on the voltage level (current) on the sp8266 if i have a free flow of water or it the tap is closed.

From the duty factor i can calculate the power its consumption during 5 minuntes and the of startup are 10 or more. Than I know if pressure vessel needs more air or is broken, or if i have a leak somewhere.

The INA219 is not a option now. The wires are between 6-8 awg and 230/400 Vac ( three phase ) upto 32 Amps. https://en.wikipedia.org/wiki/Current_transformer

Maybe I can connect the INA219 directly to the current transformer. Next step.

If i could average the measurements with tasmota ( a 5 second average), maybe i could start to make my own pump controller. The one thing i'm missing is a pressure sensor.

**Comment by Damave** [comment]
see: https://en.wikipedia.org/wiki/Current_transformer if you have a resistor in the secondary loop, you can measure the voltage with the ESP8266. And even make a 0/1 out of it by software. No aligning of a hall-sensor...

**Comment by barbudor** [comment]
INA219 is for DC current only
The output of a CT is alternative. Low voltage but alternative.

Tasmota support CT sensor on the ADC however I don't know what exact schematic is expected.
AC to DC rectification is problematic because of the low voltage (a few volts full scale generally) vs the diodes drop voltage.
A proprer 0V-drop rectifier would need to use OpAmps.
Then filtering.

But if your purpose is to only detect operation, and not measure energy accurately, a simple circuit to generate a digital 0/1 is enough. You still must take care of not sending negative voltage to the ESP.

The schematic below is showing an example for moving the alternative up to a medium point. Beware that this is for Arduino where ADC is 5VDC full scale while on the ESP it is 1VDC full scale (or 3.3V full scale at the input of a MiniD1/NodeMCU)

![image](https://user-images.githubusercontent.com/2996042/103519575-0dd36f80-4e76-11eb-98c9-781f472389d8.png)
(from https://learn.openenergymonitor.org/electricity-monitoring/ct-sensors/interface-with-arduino)

At this point I'm not sure exactly what help you need, you seem to have already advanced. Let us know

Best regards

**Reply by Damave** [reply]
Thanks for input.

You have also do a RMS (Root Mean Squire) to get the effective value of I. 
You can also split the burden resistor to lower the voltage. You need to samplerate at least with 10 times the 50 Hz of 60 Hz.
2 times sampling of the frequency can also be at the zero crossing, so this wil have a error of 100 %.
10 times oversampling will give 5 measurements of sin [0] to sin [Pi].... stil not perfect....
 
![image](https://user-images.githubusercontent.com/18482399/103575033-494f5780-4ed1-11eb-974e-1fcdc920e8fa.png)
![image](https://user-images.githubusercontent.com/18482399/103575317-d09ccb00-4ed1-11eb-935b-a7906c077f6d.png)
Source : Art of Electronics ed. 3

It should work for small signals.  There is a single sided rectifier version also.
Apply a low pass filter with a RC value of 0.1 .. 0.3 to get a maximum value and multiply this value by 0.707. 
The sample rate can now be around 1 to 10 Hz.

**Comment by lalo-uy** [comment]
If on/off is all you want just rectify the transformer output and tour on a
transistor. That will go to ant switch input.

**Reply by Damave** [reply]
That was the first idea, but now i know also the pressure. I see a relationship between pressure and current.

**Comment by barbudor** [comment]
@Damave when your project is completed, do not hesitate to write a little article in the Show and Tell section 👍

**Reply by Damave** [reply]
Some more information.
Lesson 1. First measure.
![image](https://user-images.githubusercontent.com/18482399/104225469-b7060100-5446-11eb-97e5-f81693fa6fc8.png)
This is under load (Yellow is the voltage and blue is the current).

Lesson 2. Measure without load.
![image](https://user-images.githubusercontent.com/18482399/104226085-8c687800-5447-11eb-9408-659076f1ed86.png)

How can I get the noise out in tasmota?

Will do, tomorrow, a measurement with a resistor load.

**Reply by barbudor** [reply]
Can you explain what we see on those screenshots ?
I see that the blue "noisy" track on the 2nd has a scale of 1/10th of the 1st. So if bot were at the same scale, the 2nd would be much smaller (less than 1/10th of the amplitude of the 1st screenshot)

**Reply by Damave** [reply]
Hi,

See first picture is 
![image](https://user-images.githubusercontent.com/18482399/104233792-c5f2b080-5452-11eb-886b-455caf48d69c.png)
and the second picture 
![image](https://user-images.githubusercontent.com/18482399/104233867-e28ee880-5452-11eb-8ce5-8e829865a344.png)
 are the settings of the yellow channel and the blue channel 500 mv is 500 mv/division
Op 11-01-2021 om 20:47 schreef Barbudor:
>
> Can you explain what we see on those screenshots ?
> I see that the blue "noisy" track on the 2nd has a scale of 1/10th of the 1st. So if bot were at the same scale, the 2nd would be much smaller (less than 1/10th of the amplitude of the 1st screenshot)
The noise channel is amplified by 10 times, the see the noise better.

But this is the current Transformer (CT) through 10kOhm, Normal i measure it with a resistor of 1kOhm.

What you can see it the spike of current in the first picture, and not a sine wave. And the spike is just a few degree after the 0 volt crossing. That means that the Cos phi is .... 80 degree or no nearly any power consumption. And for 90 % the current is 0.

Tomorrow I will do a measurement with a old fashion lamp. There is snow and it's freezing.

You have 3 kinds of power at AC. Watt, VA and VAR.

See https://en.wikipedia.org/wiki/Power_factor

**Reply by Frogmore42** [reply]
Are you trying to use the ADC on the esp8266 to measure power consumption of your well pump?
If so, you will be disappointed.  The ADC of the esp8266 is not specified to ANY accuracy.  You would be best to avoid using it for anything you want to measure with any accuracy.

As far as current transformer/sensors go, they are designed to push current into VERY small resistances.  Normally this would be in the 10 to 20 ohm range.  You can read more about current transformers here: https://testguy.net/content/190-Current-Transformer-Basics-Understanding-Ratio-Polarity-and-Class 
Look at Maximum Burden for a description of this.

If you want something that can run Tasmota, this might work: 
https://shopusa.shelly.cloud/shelly-em-50a-clamp-wifi-smart-home-automation#202

If you aren't looking to measure current accurately, I would hook up the current sensor to an led (two back to back, so one works no matter the polarity) and have the optically coupled to a photodiode (you can use an optocoupler, some are AC capable).  That will give you a pulsing signal, which will need filtering to get something usable.

If your really want to measure power accurately, I use and recommend http://iotawatt.com/

**Reply by Damave** [reply]
> @Damave when your project is completed, do not hesitate to write a little article in the Show and Tell section +1
[AC@DC_converter.pdf](https://github.com/arendst/Tasmota/files/7363393/AC%40DC_converter.pdf)
Took some time, still tweaking the schematics. But it's going to work. Uin = 5 Volt and voltage output set with 50 k potmeter.

**Reply by Damave** [reply]
> @Damave when your project is completed, do not hesitate to write a little article in the Show and Tell section +1

There was a fault in the schematics...
I use now a wemos D1 mini pro with a 3v3 max ADC input signal and 5Volt....
The AZ-1000 can measure upto 50 Amp but linear till 30 Amp.
 With RV1 you can change the output voltage.

[AC2DC_converter.pdf](https://github.com/arendst/Tasmota/files/7364024/AC2DC_converter.pdf)
..

**Comment by Frogmore42** [comment]
So what are you really trying to do?
At the start, you said you only cared about the time the pump was on.  If that is still the case there are some solutions for that, but also challenges.

If, instead, you really do want to measure the power that the pump is using, you are going to have to decide how accurate a measurement you want.  The greater the accuracy you want, the more it will cost and the less able you will be to achieve that with a standard esp8266.

The ADC in the esp8266 is not specified to any accuracy, so it is a poor choice to use for measuring anything that needs any kind of precision/accuracy.  That does not mean the esp8266 is bad, just that I wouldn't trust it with anything that I cared about.

You said you had a 3 phase motor in your pump and that it was using a motor controller.  The one you linked didn't really indicated how it worked.  There are two main methods for this.  First is a variable frequency drive system.  These are very expensive and don't seem to have a very high reliability.  They change the frequency of the power applied to the 3ph motor which causes it change its speed.  This changes the pump curve (flow vs pressure) significantly.  Since it is using an electronic system to do this, it probably has a non-linear current demand and will require a sophisticated circuit to accurately measure power.  It all depends on what accuracy level you need/want.  if +/- 10-20% is acceptable to you, you can probably create something on your own that will achieve that, but even that won't be easy.

If, instead, you have a standard 3phase motor that uses a back pressure valve to limit flow to keep the pump on longer, then you can probably measure a single leg of the motor with something like this: https://shop.shelly.cloud/shelly-em-50a-clamp-wifi-smart-home-automation-1#139 
I think Tasmota will work with it, but not 100% sure

If you just want to see the pump is on or not, I would use this: https://shelly.cloud/knowledge-base/devices/shelly-i3/#pinout
It looks like that supports Tasmota and should show up as three switches.  It should be easy to get this in to Domoticz

Now, if you REALLY, REALLY, REALLY want to create some power measurement circuitry, I would suggest spending a lot of time reviewing  the information here: https://learn.openenergymonitor.org/electricity-monitoring/ac-power-theory/introduction
It is a great introduction into AC power theory and practical ways to measure it.

If you really just want to measure your whole house consumption, I use and recommend this: http://iotawatt.com/
A lot of people use it with 3ph power and it does a good job and is quite accurate.
